packages: build zlib and zlib-devel

arnaldo2792 opened the pull request:
**Issue number:**
N / A

**Description of changes:**
```
6bd13fa5 packages: build zlib and zlib-devel
```

This commit adds zlib runtime libraries and development headers

**Testing done:**
- Build vmware-dev variant and confirm the new libraries are in `/usr/lib`, the snapshot is the output of a VMWare host:

![image](https://user-images.githubusercontent.com/6305870/116148034-4604f500-a695-11eb-92a5-de74cfaf4f37.png)

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

arnaldo2792 committed:
packages: build zlib and zlib-devel

This commit adds zlib runtime libraries and development headers

bcressey commented on packages/libz/libz.spec:
```diff
@@ -0,0 +1,60 @@
+%make_build
+
+%check
+make test
```
We don't include `%check` scriptlets since in cross-compiling scenarios, we're not guaranteed to be able to run the programs we've built.

tjkirch commented on variants/vmware-dev/Cargo.toml:
```diff
@@ -24,6 +24,8 @@ included-packages = [
     "docker-proxy",
 # tools
     "findutils",
+    # FIXME: remove this when open-vm-tools is packaged
```
I'm not sure I understand this - if we don't need zlib for something in the image yet, we can just remove this; if it's needed for testing, that's fine, but we wouldn't want to merge it, so I think we can add it locally but leave it out of the PR.

bcressey commented on packages/libz/libz.spec:
```diff
@@ -0,0 +1,60 @@
+%set_env
+%make_install
+# We don't want the static libraries
+rm %{buildroot}/%{_cross_libdir}/libz.a
```
Typically we include static libraries in the `devel` subpackage, even if we don't expect them to be used.

tjkirch commented on packages/libz/Cargo.toml:
```diff
@@ -0,0 +1,16 @@
+
+[[package.metadata.build-package.external-files]]
+url = "http://www.zlib.net/zlib-1.2.11.tar.gz"
+sha512 = "73fd3fff4adeccd4894084c15ddac89890cd10ef105dd5e1835e1e9bbb6a49ff229713bd197d203edfa17c2727700fce65a2a235f07568212d820dca88b528ae"
```
```suggestion
url = "https://zlib.net/zlib-1.2.11.tar.xz"
sha512 = "b7f50ada138c7f93eb7eb1631efccd1d9f03a5e77b6c13c8b757017b2d462e19d2d3e01c50fad60a4ae1bc86d431f6f94c72c11ff410c25121e571953017cb67"
```

tjkirch commented on packages/libz/libz.spec:
```diff
@@ -0,0 +1,60 @@
+URL: https://www.zlib.net/
+License: zlib and Boost
+
+Source: http://www.zlib.net/zlib-%{version}.tar.gz
```
```suggestion
Source: http://www.zlib.net/zlib-%{version}.tar.xz
```

bcressey commented on packages/libz/libz.spec:
```diff
@@ -0,0 +1,60 @@
+Release: 1%{?dist}
+Summary: Compression and decompression library
+URL: https://www.zlib.net/
+License: zlib and Boost
```
This needs to be the SPDX license identifier, so `Zlib` in this case:
https://spdx.org/licenses/Zlib.html

I'm not sure we ship anything that requires us to include the Boost license.

tjkirch replied:
It looks like the boost license only applies to some contrib code, dotzlib, which I suspect we don't use.  Do we need that?

bcressey commented on packages/libz/libz.spec:
```diff
@@ -0,0 +1,60 @@
+Summary: Compression and decompression library
+URL: https://www.zlib.net/
+License: zlib and Boost
+
+Source: http://www.zlib.net/zlib-%{version}.tar.gz
+
```
nit: I'd remove these empty lines

bcressey commented on packages/libz/libz.spec:
```diff
@@ -0,0 +1,60 @@
+%{summary}.
+
+%package devel
+Summary: Header files and libraries for Zlib development
```
My convention for these summaries is:
"Summary: Files for development using the $main_package_description"

```suggestion
Summary: Files for development using the library for zlib compression 
```

bcressey commented on packages/libz/libz.spec:
```diff
@@ -0,0 +1,60 @@
+Name: %{_cross_os}libz
+Version: 1.2.11
+Release: 1%{?dist}
+Summary: Compression and decompression library
```
nit:
```suggestion
Summary: Library for zlib compression
```

tjkirch commented on packages/libz/libz.spec:
```diff
@@ -0,0 +1,60 @@
+export CC="%{_cross_target}-gcc" \\\
+export DESTDIR='%{buildroot}' \\\
+export PREFIX='%{_cross_prefix}' \\\
+export SHLIBDIR='%{_cross_libdir}' \\\
```
I'm not seeing any reference to SHLIBDIR in the source - what does this do?

bcressey commented on packages/libz/libz.spec:
```diff
@@ -0,0 +1,60 @@
+# We don't want the static libraries
+rm %{buildroot}/%{_cross_libdir}/libz.a
+
+%files
+%license README
+%{_cross_attribution_file}
+%{_cross_libdir}/libz.so.*
+%exclude %{_cross_mandir}
+
+%files devel
+%{_cross_libdir}/libz.so
+%{_cross_libdir}/pkgconfig/zlib.pc
+%{_cross_includedir}/zlib.h
+%{_cross_includedir}/zconf.h
```
nit: I tend to use boilerplate for these includes, rather than package-specific ones:
```suggestion
%files
%license README
%{_cross_attribution_file}
%{_cross_libdir}/*.so.*
%exclude %{_cross_mandir}

%files devel
%{_cross_libdir}/*.so
%{_cross_includedir}/*.h
%{_cross_pkgconfigdir}/*.pc
```

bcressey commented on packages/libz/libz.spec:
```diff
@@ -0,0 +1,60 @@
+%set_cross_build_flags \\\
+export CC="%{_cross_target}-gcc" \\\
+export DESTDIR='%{buildroot}' \\\
+export PREFIX='%{_cross_prefix}' \\\
```
zlib also supports `CROSS_PREFIX` which seems like it'd be more correct to use

tjkirch replied:
I don't see this being used in the zlib build, just the `--prefix` argument you have below - is this needed?  We could pass `%{_cross_prefix}` below.

bcressey commented on packages/libz/libz.spec:
```diff
@@ -0,0 +1,60 @@
+# required to `make install` libz
+%global set_env \
+%set_cross_build_flags \\\
+export CC="%{_cross_target}-gcc" \\\
```
zlib supports `CROSS_PREFIX` which seems like it'd be more correct to use to help find the right toolchain

tjkirch commented on packages/libz/libz.spec:
```diff
@@ -0,0 +1,60 @@
+
+%build
+%set_env
+# Force --prefix=$PREFIX
```
```suggestion
# zlib only reads prefix from this argument, not the environment.
```

arnaldo2792 commented:
- Addressed all the nits
- Removed Boost from the licenses list
- Removed unused environment variables
- Switch to use the `xz` file instead of `gz`

arnaldo2792 commented:
- Re-added `set_env` macro to setup environment variables
- Removed dependency from vmware-dev variant
- Removed Boost license comment

tjkirch commented on variants/vmware-dev/Cargo.toml:
```diff
@@ -1,5 +1,4 @@
-[package]
-name = "vmware-dev"
+[package] name = "vmware-dev"
```
This change snuck in :)

tjkirch commented on packages/libz/libz.spec:
```diff
@@ -0,0 +1,52 @@
+
+%install
+%set_env
+make install
```
```suggestion
%make_install
```

tjkirch commented on packages/libz/libz.spec:
```diff
@@ -0,0 +1,52 @@
+
+%build
+%set_env
+# zlib supports $CROSS_PREFIX to setup the targeted compiler, and
+# it only reads prefix from this argument, no the environment
```
```suggestion
# zlib only reads prefix from this argument, not the environment.
```

arnaldo2792 commented:
- Use the `%make_install` macro
- Remove comment explaining the use of `CROSS_PREFIX`
- Fixed funky `Cargo.toml` formatting

bcressey commented on packages/libz/libz.spec:
```diff
@@ -0,0 +1,51 @@
+
+%build
+%set_env
+# zlib only reads prefix from this argument, no the environment
```
nit:
```suggestion
# zlib only reads prefix from this argument, not the environment
```

bcressey commented on packages/libz/libz.spec:
```diff
@@ -0,0 +1,51 @@
+%global set_env \
+%set_cross_build_flags \\\
+export CROSS_PREFIX="%{_cross_target}-" \\\
+export DESTDIR='%{buildroot}' \\\
```
nit: this should be set by `%make_install`
```suggestion
```

arnaldo2792 commented:
- Remove unnecessary `DESTDIR=`
- Fix typo in comment
